### PR TITLE
PSP Client

### PR DESCRIPTION
**Pull request by Magnus93**
- [x] https://github.com/payfunc/model-base/pull/33

### CODE REVIEW

File: Client/Customer.ts [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+		return await this.connection.post<Customer.Method.Card>(`customer/${customer}/method`, request)
+	}
+	async setMethods(request: Customer.Method, customer: string): Promise<Customer.Method[] | gracely.Error> {
+		return await this.connection.put<Customer.Method[]>(`customer/${customer}/methods`, request)
+	}
+	async createSubscription(request: Subscription.Creatable, customer: string): Promise<Subscription[] | gracely.Error> {
+		return await this.connection.post<Subscription[]>(`customer/${customer}/subscription`, request)
+	}
+	async cancelSubscription(customer: string, subscription: string): Promise<Subscription[] | gracely.Error> {
+		return await this.connection.remove<undefined>(`customer${customer}/subscription/${subscription}`)
```
Comment (nilssonemma) [inline Client/Customer.ts:51]: You forgot the "/" after customer (also change undefined to the correct return type)

File: Client/Customer.ts [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import { Customer } from "../Customer"
+import { Item } from "../Item"
+import { Customer as PaymentCustomer } from "../Payment/Customer"
+import { Subscription } from "../Subscription"
+
+export class CustomerClient {
+	constructor(readonly connection: Connection) {}
+
+	async balanceItems(request: number | Item | Item[], customer: string): Promise<gracely.Result | gracely.Error> {
+		return await this.connection.post<gracely.Result>(`customer/${customer}/balance`, request)
```
Comment (nilssonemma) [inline Client/Customer.ts:12]: Should be await this.connection.post<gracely.Result | gracely.Error>
Add gracely.Error in the functions below too

**Review by nilssonemma — CHANGES_REQUESTED** [review]

File: Client/Customer.ts (hunk 2 above)
Comment (Magnus93) [reply Client/Customer.ts:11]: Nope, specifying gracely.Error is not necessary
https://github.com/payfunc/model-base/blob/master/Connection.ts#L33

File: Client/Customer.ts [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+		return await this.connection.get<ModelCustomer>(`customer/${customer}`)
+	}
+	async update(request: ModelCustomer, customer: string): Promise<ModelCustomer | gracely.Error> {
+		return await this.connection.patch<ModelCustomer>(`customer/${customer}`, request)
+	}
+	async list(): Promise<ModelCustomer[] | gracely.Error> {
+		return await this.connection.get<ModelCustomer[]>("customer/")
+	}
+	async createKey(customer: string): Promise<string | gracely.Error> {
+		return await this.connection.post<string>(`customer/${customer}`, undefined)
```
Comment (marcushanses) [inline Client/Customer.ts:29]: why undefined?

File: Client/Customer.ts [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+	}
+	async createLink(customer: string): Promise<ModelCustomer.Link | gracely.Error> {
+		return await this.connection.post<ModelCustomer.Link>(`customer/${customer}/link`, undefined)
```
Comment (marcushanses) [inline Client/Customer.ts:32]: why undefined?

**Review by marcushanses — APPROVED** [review]

**Review by nilssonemma — APPROVED** [review]

**Review by fredrikbryntesson — APPROVED** [review]

**Pull request merged**